src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,206 @@
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexOutOfBounds(index);
+        return unlink(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (x.value == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (object.equals(x.value)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkEmptyList(T value) {
+        Node<T> newNode = new Node<>(value);
+        head = newNode;
+        tail = newNode;
+        size++;
+    }
+
+    private void linkHead(T value) {
+        final Node<T> f = head;
+        final Node<T> newNode = new Node<>(value, f);
+        head = newNode;
+        f.prev = newNode;
+        size++;
+    }
+
+    private void linkTail(T value) {
+        Node<T> l = tail;
+        Node<T> newNode = new Node<>(l, value);
+        tail = newNode;
+        l.next = newNode;
+        size++;
+    }
+
+    private Node<T> getNode(int index) {
+        if (index < size / 2) {
+            Node<T> x = head;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x;
+        }
+        Node<T> x = tail;
+        for (int i = size - 1; i > index; i--) {
+            x = x.prev;
+        }
+        return x;
+    }
+
+    private void checkIndexOutOfBounds(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("\"Index\" " + index + "\" Size\" " + size);
+        }
+    }
+
+    private void checkForEmptyList() {
+        if (isEmpty()) {
+            throw new NullPointerException("Can't unlinked from from an empty list");
+        }
+    }
+
+    private T unlink(Node<T> node) {
+        checkForEmptyList();
+        final T element = node.value;
+        final Node<T> next = node.next;
+        final Node<T> prev = node.prev;
+        if (next == null && prev == null) {
+            size = 0;
+            return element;
+        }
+        if (prev == null) {
+            head = next;
+            next.prev = head;
+            size--;
+            return element;
+        }
+        if (next == null) {
+            tail = prev;
+            prev.next = tail;
+            size--;
+            return element;
+        }
+        next.prev = prev;
+        prev.next = next;
+        size--;
+        return element;
```

Comment:
Considering all these comments, we will have this code:
```suggestion
        if (prev == null) {
            head = next;
        } else if (next == null) {
            tail = prev;
        } else {
            next.prev = prev;
            prev.next = next;
        }
        size--;
        return element;
```